actionview/lib/action_view/renderer/collection_renderer.rb

```diff
@@ -109,39 +109,51 @@ def each_with_info
       end
     end
 
+    def render_collection_derive_partial(collection, context, block)
+      paths = collection.map { |o| partial_path(o, context) }
+
+      if paths.uniq.length == 1
+        # Homogeneous
+        render_collection_with_partial(collection, paths.first, context, block)
+      else
+        # Heterogeneous
+        if @options[:cached]
+          raise NotImplementedError, "render caching requires a template. Please specify a partial when rendering"
+        end
+
+        render_collection_with_multiple_partials(collection, paths, context, block)
+      end
+    end
+
     def render_collection_with_partial(collection, partial, context, block)
       iter_vars  = retrieve_variable(partial)
 
       collection = if collection.respond_to?(:preload_associations)
-        PreloadCollectionIterator.new(collection, partial, iter_vars, collection)
-      else
-        SameCollectionIterator.new(collection, partial, iter_vars)
-      end
+                     PreloadCollectionIterator.new(collection, partial, iter_vars, collection)
+                   else
+                     SameCollectionIterator.new(collection, partial, iter_vars)
+                   end
 
       template = find_template(partial, @locals.keys + iter_vars)
 
       layout = if !block && (layout = @options[:layout])
-        find_template(layout.to_s, @locals.keys + iter_vars)
-      end
+                 find_template(layout.to_s, @locals.keys + iter_vars)
+               end
```

Comment:
Similar, please don't update indentation in unrelated lines of code.

Reply:
Thanks @p8, it should be fixed now.